File: linked_list/203_remove_linked_list_elements.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        curr, prev = head, None

        while curr:
            if curr.val == val:
                if curr == head:
                    head = head.next
                    curr = head
                else:
                    prev.next = curr.next
                    curr = curr.next
            else:
                prev = curr
                curr = curr.next

        return head
        

        
        

File: linked_list/203_remove_linked_list_elements.java
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        // head = [1,2,6,3,4,5,6], val = 6 -> [1,2,3,4,5]
        ListNode res = new ListNode(0);
        ListNode prev = res;
        
        while (head != null) {
            if (head.val != val) {
                prev.next = head;
                prev = prev.next;
            }
            head = head.next;
        }
        prev.next = null;
        return res.next;
        // O(N) + O(1)
    }
}